src/Nethermind/Nethermind.Network.Discovery/Discv5/DiscoveryV5App.cs

```diff
@@ -0,0 +1,309 @@
+                    }
+                    if (distances[0] + i <= 256)
+                    {
+                        distances[n++] = distances[0] + i;
+                    }
+                }
+
+                return distances;
+            }
+
+            ConcurrentQueue<IEnr> nodesToCheck = new(getAllNodes);
+            HashSet<IEnr> checkedNodes = [];
+
+            while (!token.IsCancellationRequested)
+            {
+                if (!nodesToCheck.TryDequeue(out IEnr? newEntry))
+                {
+                    return;
+                }
+
+                NodeAddedByDiscovery(newEntry);
+
+                if (!checkedNodes.Add(newEntry))
+                {
+                    continue;
+                }
+
+                IEnr[]? newNodesFound = (await _discv5Protocol.SendFindNodeAsync(newEntry, GetDistances(newEntry.NodeId, nodeId)))?.Where(x => !checkedNodes.Contains(x)).ToArray();
+
+                if (newNodesFound is not null)
+                {
+                    foreach (IEnr? node in newNodesFound)
+                    {
+                        nodesToCheck.Enqueue(node);
+                    }
+                }
+            }
+        }
+
+        IEnr[] GetStartingNodes()
+        {
+            return _discv5Protocol.GetAllNodes.ToArray();
+        }
+
+        Random random = new();
+        await _discv5Protocol!.InitAsync();
+
+
+        if (_logger.IsDebug) _logger.Debug($"Initially discovered {_discv5Protocol.GetActiveNodes.Count()} active peers, {_discv5Protocol.GetAllNodes.Count()} in total.");
+
+        byte[] randomNodeId = new byte[32];
+        while (!_appShutdownSource.IsCancellationRequested)
+        {
+            try
+            {
+                await DiscoverAsync(GetStartingNodes(), _discv5Protocol.SelfEnr.NodeId, _appShutdownSource.Token);
+
+                for (int i = 0; i < 3; i++)
+                {
+                    random.NextBytes(randomNodeId);
+                    await DiscoverAsync(GetStartingNodes(), randomNodeId, _appShutdownSource.Token);
+                }
+            }
+            catch (Exception ex)
+            {
+                if (_logger.IsError) _logger.Error($"Discovery via custom random walk failed.", ex);
+            }
+
+            if (_api.PeerManager?.ActivePeers.Any() == true)
+            {
+                await Task.Delay(_discoveryConfig.DiscoveryInterval, _appShutdownSource.Token);
+            }
+        }
+    }
+
+    public async Task StopAsync()
+    {
+        if (_api.PeerManager is null)
+        {
+            return;
+        }
+
+        List<EntrySecp256K1> activeNodes = _api.PeerManager.ActivePeers.Select(x => new EntrySecp256K1(NBitcoin.Secp256k1.Context.Instance.CreatePubKey(x.Node.Id.PrefixedBytes).ToBytes(false))).ToList();
+
+        IEnumerable<IEnr> activeNodeEnrs = _discv5Protocol.GetAllNodes.Where(x => activeNodes.Any(n => n.Equals(x.GetEntry<EntrySecp256K1>(EnrEntryKey.Secp256K1))));
```

Comment:
EntrySecp256K1 doesn't overload `Equals`, so it will be only reference comparision?